### ISSUE TITLE
Iterator.scanLeft has incorrect element-by-element behavior (delayed by one item)

### ISSUE THREAD

**Issue by klaeufer** [scala/bug#10709]
# Version  info

Scala 2.12.4
JVM 1.8.0_144

# Example

```
object Main extends App {
  val lines = scala.io.Source.stdin.getLines
  val results = lines.scanLeft(0)((count, line) => count + line.length)
  results.foreach(println)
}
```

# Observed behavior

The program does not print the initial value until it reads the first line, even though this value is available immediately. Subsequent counts are printed for the previous line. The count for the very last line is printed only after EOF. 

```
$ ./target/universal/stage/bin/main
hello
0
world
5
what
10
up
14
^D
16
```

# Expected behavior

The program should immediately print the initial count. It should print subsequent counts right after reading each line. It terminates upon EOF (without printing anything else).

```
$ ./target/universal/stage/bin/main
0
hello
5
world
10
what
14
up
16
^D
```

# Analysis

`Iterator.scanLeft` holds back one item in its internal state because of the premature call to `self.next()` on this line: https://github.com/scala/scala/blob/v2.12.4/src/library/scala/collection/Iterator.scala#L600

# Proposed fix

This example contains two implementations, a high-level one using `iterate`, `takeWhile`, and `map`; and a low-level one based on fusing `++`, `iterate`, and `takeWhile`.

https://github.com/LoyolaChicagoCode/iterators-scala/blob/master/src/main/scala/functional/modular/CumulativeLengthFunctionalModular.scala

**Comment by som-snytt** [comment]
Many of the unit tests are tests of laziness.

**Comment by klaeufer** [comment]
Thanks a lot, this has been fixed as of 2.12.5. Is someone from the Scala library team going to close the issue?
